lib/src/rules/dangling_library_doc_comments.dart

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Attach library doc comments to library directives.';
+
+const _details = r'''
+Attach library doc comments (with `///`) to library directives, rather than
+leaving them dangling near the top of a library.
+
+**BAD:**
+```dart
+/// This is a great library.
+import 'package:math';
+```
+
+```dart
+/// This is a great library.
+
+class C {}
+```
+
+**GOOD:**
+```dart
+/// This is a great library.
+library;
+
+import 'package:math';
+
+class C {}
+```
+
+**NOTE:** An unnamed library, like `library;` above, is only supported in Dart
+2.19 and later. Code which might run in earlier versions of Dart need to provide
+a name in the `library` directive.
+''';
+
+class DanglingLibraryDocComments extends LintRule {
+  static const LintCode code = LintCode(
+      'dangling_library_doc_comments', 'Dangling library doc comment.',
+      correctionMessage: 'Attach library doc comments to library directives.');
+
+  DanglingLibraryDocComments()
+      : super(
+            name: 'dangling_library_doc_comments',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  LintCode get lintCode => code;
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this);
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final DanglingLibraryDocComments rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    if (node.directives.isNotEmpty) {
+      // Only consider a doc comment on the first directive. Doc comments on
+      // other directives do not have the appearance of documenting the library.
+      var firstDirective = node.directives.first;
+      if (firstDirective is LibraryDirective) {
+        // Given the presense of library directive, don't worry about later doc
+        // comments in the library.
+        return;
+      }
+      if (firstDirective is PartOfDirective) {
+        // Don't worry about possible "library doc comments" in a part.
+        return;
+      }
+
+      if (firstDirective.documentationComment != null) {
+        rule.reportLint(firstDirective.documentationComment);
+        return;
+      }
+    }
+
+    if (node.declarations.isEmpty) {
```

Comment:
I think this needs to be an else-if so that we don't get here if there are directives.

Reply:
Hmm, I don't think so. I can see someone writing a supposed library dart doc between the directives and the declarations.

Reply:
Given

```
import 'dart:async';

/// Comment

class C {}
```

I think that we shouldn't lint the doc comment. Neither analyzer nor dartdoc would have previously interpreted this as the doc comment for the library, so I don't think this lint should either.

Reply:
That's too bad. I'm sure will miss some cases. Done, in any case.